sktime/annotation/hmm.py

```diff
@@ -0,0 +1,275 @@
+                paths[:, j] += np.log(emi_probs[:, i])  # adds log(probs_sub) row-wise
+            paths += np.log(
+                transition_prob_mat
+            )  # adds log(transition_prob_mat) element-wise
+            trans_id[:, i] = np.argmax(paths, axis=0)
+            trans_prob[:, i] = np.max(paths, axis=0)
+
+        if np.any(np.isinf(trans_prob[:, -1])):
+            raise ValueError("Change parameters, the distribution doesn't work")
+
+        return trans_prob, trans_id
+
+    @classmethod
+    def _make_emission_probs(
+        cls, emission_funcs: list, observations: np.ndarray
+    ) -> np.ndarray:
+        """Calculate the prob each obs comes from each hidden state."""
+        # assign emission probabilities from each state to each position:
+
+        emi_probs = np.zeros(shape=(len(emission_funcs), len(observations)))
+        for state_id, emission_tuple in enumerate(emission_funcs):
+            emission_func = emission_tuple[0]
+            kwargs = emission_tuple[1]
+            emi_probs[state_id, :] = np.array(
+                [emission_func(x, **kwargs) for x in observations]
+            )
+        return emi_probs
+
+    def _hmm_viterbi_label(self) -> np.array:
+        """Assign hidden state ids to all observations based on most likely path.
+
+        Parameters
+        ----------
+        self - an HMM instance which already has already
+
+        Returns
+        -------
+            -hmm_fit: a 2xn array with the first column being position and the second
+                column being a peak assignment.
+        """
+        hmm_fit = np.zeros(self.num_obs)
+        # Now we trace backwards and find the most likely path:
+        max_inds = np.zeros(self.num_obs, dtype=np.int32)
+        max_inds[-1] = np.argmax(self.trans_prob[:, -1])
+        hmm_fit[-1] = self.states[max_inds[-1]]
+        for index in reversed(list(range(1, self.num_obs))):
+            max_inds[index - 1] = self.trans_id[max_inds[index], index]
+            hmm_fit[index - 1] = self.states[max_inds[index - 1]]
+        return hmm_fit
+
+    def _fit(self, X, Y=None):
+        """Calculate the transition matrices for the given model/observation."""
+        self.num_obs = len(X)
+        emi_probs = HMM._make_emission_probs(self.emission_funcs, X)
+        init_probs = self.initial_probs
+        # if no initial_probs were supplied assign all states equal prob:
+        if self.initial_probs is None:
+            init_probs = 1.0 / (self.num_states) * np.ones(self.num_states)
+        trans_prob, trans_id = HMM._calculate_trans_mats(
+            init_probs,
+            emi_probs,
+            self.transition_prob_mat,
+            self.num_obs,
+            self.num_states,
+        )
+        self.trans_prob = trans_prob
+        self.trans_id = trans_id
+        return self
+
+    def _predict(self, X):
+        """Use the transition matrices to calculate the most likely state."""
+        return self._hmm_viterbi_label()
```

Comment:
this can´t be right - `_hmm_viterbi_label` not using `X`? Can this only be used in-sample?

Reply:
Well, as you point out below the estimator (as it is) is only applying Viterbi to a single observation (and originally all the logic that extracted information from `X` was in `_fit`.  I have since moved everything to `_predict`.